pkg/forwarder/forwarder.go

```diff
@@ -368,5 +385,72 @@ func (w *Worker) forward(ctx context.Context) error {
 			rlogger.Log(w.logger, rlogger.Warn, "msg", failedStatusReportMsg, "err", err)
 		}
 	}
+
+	elapsed := time.Since(fStart)
+	rlogger.Log(w.logger, rlogger.Info, "forward_total_time", elapsed)
 	return err
 }
+
+func (w *Worker) getFederateMetrics(ctx context.Context) ([]*clientmodel.MetricFamily, error) {
+	start := time.Now()
+	from := w.from
+
+	// reset query from last invocation, otherwise match rules will be appended
+	w.from.RawQuery = ""
+	v := from.Query()
+	for _, rule := range w.rules {
+		v.Add("match[]", rule)
+	}
+	from.RawQuery = v.Encode()
+
+	req := &http.Request{Method: "GET", URL: from}
+	families, err := w.fromClient.Retrieve(ctx, req)
+	if err != nil {
+		return nil, err
+	}
+
+	elapsed := time.Since(start)
+	rlogger.Log(w.logger, rlogger.Info, "federate_time", elapsed)
+	return families, nil
+}
+
+func (w *Worker) getRecordingMetrics(ctx context.Context) ([]*clientmodel.MetricFamily, error) {
+	families := []*clientmodel.MetricFamily{}
+	start := time.Now()
+	from := w.from
+
+	originPath := from.Path
+	from.Path = "/api/v1/query"
+	defer func() {
+		w.from.Path = originPath
+	}()
+
+	for _, rule := range w.recordingRules {
+		var r map[string]string
+		err := json.Unmarshal(([]byte)(rule), &r)
+		if err != nil {
+			rlogger.Log(w.logger, rlogger.Warn, "msg", "Input error", "err", err)
+			continue
+		}
+		rname := r["name"]
+		rquery := r["query"]
+
+		// reset query from last invocation, otherwise match rules will be appended
+		from.RawQuery = ""
+		v := w.from.Query()
+		v.Add("query", rquery)
+		from.RawQuery = v.Encode()
+
+		req := &http.Request{Method: "GET", URL: from}
+		rfamilies, err := w.fromClient.RetrievRecordingMetrics(ctx, req, rname)
+		if err != nil {
+			return families, err
```

Comment:
if err, I think use "continue" is better

Reply:
fixed, will only print warning and continue